### ISSUE TITLE
[RAG meta-data] Include meta-data of the RAG retrieved documents

### ISSUE THREAD

**Issue by coreation** [pinecone-io/canopy#186]
### Is this your first time submitting a feature request?

- [X] I have searched the existing issues, and I could not find an existing issue for this feature
- [X] I am requesting a straightforward extension of existing functionality

### Describe the feature

As far as I can see, for the RAG generated response, there's no meta-data with regards to which vector documents it has used. This would be very useful for end-users so that they can get references as to which documents the response is referring to.

**Edit** Having dived into the code base, is this intended to be build perhaps by clients using this framework by building their own ContextBuilder?

### Who will this benefit?

I believe there's a lot use cases that would want to prove that what they output is correct. The end user can manually verify or get more in-depth information based on meta-data coming back from RAG retrieved documents.

### Are you interested in contributing this feature?

My Python knowledge is very limited, but I'll give it a go if maintainers direct me as to how they see the data object structure change for example.

### Anything else?

Nothing at the moment, I think the feature is pretty straight-forward. I'd suggest the client supports a simple list of strings representing the meta-data properties they would like to retrieve from Pinecone, and include those in the RAG response.

**Comment by miararoy** [comment]
hi @coreation ! 

Thanks for the feedback and the suggestion. 

First, Canopy supports this explicitly by referring to the `source` field on the upserted documents.
Your issue indicated this is not properly documented, so I added explicit documentation on the main readme.
see PR #187 

Secondly, we have thought deeply about the idea of adding metadata in the RAG process, but have chosen to simplify things and use a single field for now. In the future we may extend this capability. 

Does this solved your original query? in any case we would love to hear more suggestions.

**Comment by coreation** [comment]
Hi @miararoy thanks so much for elaborate reply! Let me start by saying this kind of replies on open source repositories really gives a boost of confidence to get started and contribute however little I can.

1/ 
With regards to my original question and your follow-up. 

For me personally to only support the "source" property instead of all meta-data to be contained in the ContextEngine is sufficient. Your answer does lead to a follow-up of my own if you don't mind.

Looking at the [high level architecture](https://github.com/pinecone-io/canopy/blob/main/docs/library.md), there are 3 components, where - to my understanding - the source of the RAG results is available within the ContextEngine.  The ChatEngine leverages the ContextEngine as can be seen in app.py to then formulate a ChatResponse. 

However, the ChatResponse does not contain these references, but I could be wrong. 
This assumption comes from my limited Python knowledge and looking at how the DTO "[ChatResponse" is created out](https://github.com/pinecone-io/canopy/blob/main/src/canopy_server/app.py#L133C31-L133C31) of something seemingly untyped, casted [from "run_in_threadpool"](https://github.com/pinecone-io/canopy/blob/main/src/canopy_server/app.py#L117). 

In other words, if I wanted to use Canopy to make a chatbot where I use the "chat" endpoint to also include references to the RAG results, I'd have to tweak that myself...or?


2/ 
I've currently made a Q&A chatbot using Pinecone and LangChain, but finding this really answered an itch that kept coming back: why do I need something so configurable combined with so many possible features to just implement a "simple" Q&A chatbot based on my own vector data. Am I correct to assume that's the itch you felt at Pinecone as well and made this repository? By the way, huge fan of how you guys made it so damn easy to just get started with blazing fast search at Pinecone. Well made product! I hope this kudos and context gives a better idea on how I look at this repo :)

**Comment by miararoy** [comment]
Thanks for this feedback and I will pass it on to the engineering team if you don't mind. 

regarding your questions:

I'll start with 2/
Yes, this was the intuition we had, build a strong and simple RAG framework, focused around the data, or the R in RAG. We also wanted to build something robust and configureable (in a different way) so stay tuned on more learning materials coming out. 

re 1/ 
> if I wanted to use Canopy to make a chatbot where I use the "chat" endpoint to also include references to the RAG results, I'd have to tweak that myself...or?

you just need to provide the source in the source field and that is it. i recommend you test it and see. 

to understand the internals more, when ContextEngine is passing on the Context to the ChatEngine it will create "snippets" which are pieces of context, along with their source, this is being done in ContextBuilder ![for example here](https://github.com/pinecone-io/canopy/blob/main/src/canopy/context_engine/context_builder/stuffing.py#L14)

in theory this could be extended to more complex data structures in the future but at the moment we would like to keep it simple

does that make sense?

**Comment by coreation** [comment]
Hey @miararoy it does, I'll just have to set it up and see for myself and you point out :) 

re 2/ A huge 👍 for that! 

I'm looking forward to future developments and contribute where I can.

**Comment by igiloh-pinecone** [comment]
@coreation as you correctly edited in your original issue, this functionality is fully supported by users implementing their own `ContextBuilder` class.

Once you have implemented the relevant class, you can configure your Canopy server by creating a config file in [this format](https://github.com/pinecone-io/canopy/blob/main/config/config.yaml) and changing the `context_builder: type` field to point to your new class.  
Now you can run Canopy with `canopy start --config my_config.yaml` and it will implement the behavior you desire.

**Comment by coreation** [comment]
Good stuff, so the intended behaviour is people clone the repo and adjust the code where applicable, stay clear from any canopy dedicated folders as that will give merge conflicts when updating to a newer version of canopy? 

I'm just trying to wrap my head around it coming from Laravel for example. There they provide a boilerplate to get started, but the real framework is added as a dependency. So that you can change whatever you need (i.e. route endpoints) but you can still upgrade without any issue because the business logic of the framework is in a PHP package.

**Comment by coreation** [comment]
> > thanks @igiloh-pinecone , I'm assuming you can then tweak the output of the "chat" endpoint as well, adding the "source" meta-data to the output, next to the generated chat (text) response for example.
> 
> Sure, if you look at the code of the `/chat/completion` endpoint [here](https://github.com/pinecone-io/canopy/blob/63c2da4fcea578f3e191c9a1b5d1b707091a804e/src/canopy_server/app.py#L117), it calls the `ChatEngine` and gets an `answer`. That `answer` has a hidden `answer.debug_info` field that includes the full context prompt (output of `ContextBuilder`) and also the full retrieved text chunks. You can modify the code to print \ log \ return any of these fields.

@igiloh-pinecone I've added a [bug report](https://github.com/pinecone-io/canopy/issues/297) for this, it doesn't seem to add any information. I'm going to dig deeper, see if I can find the issue myself, but the CE_DEBUG_INFO flag is not doing it's job it seems.
